### PR TITLE
[grape] add Grape integration for API endpoints

### PR DESCRIPTION
**Pull request by palazzem**
### What it does

Add Grape auto instrumentation using the `Patcher` method. Grape classes are patched so that new instrumentation is available before the task is executed. It requires `ActiveSupport` and relies on Grape signals to:
* trace the whole endpoint execution
* trace `before` and `after` filters
* trace the `Endpoint` body

This auto instrumentation works standalone, or using the Rack and/or Rails auto instrumentation. If Rails is used, both Rack and Grape integrations are activated.

#### What's missing
* [x] documentation
* [ ] integration test and instrumentation for Rails

### CODE REVIEW

File: lib/ddtrace/contrib/grape/endpoint.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+          end
+
+          # catch thrown exceptions
+          span.set_error(payload[:exception_object]) unless payload[:exception_object].nil?
+
+          # ovverride the current span with this notification values
+          span.start_time = start
+          span.set_tag('grape.route.endpoint', api_view)
+          span.set_tag('grape.route.path', path)
+          span.finish_at(finish)
```
Comment (ufoot) [inline lib/ddtrace/contrib/grape/endpoint.rb]: This is globally fine. I'm nitpicking but would still probably harden it like in https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/contrib/rails/action_controller.rb#L71 by adding an `ensure` clause around `start_time = start` and `finish`. Just, to make sure it makes its way down the pipeline. The heuristic I use here is quite dumb -> if there are at least 10 lines of code, however simple they mike look, the probability it yields an exception "some day in the future" is high.

File: lib/ddtrace/contrib/grape/endpoint.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+          tracer = pin.tracer
+          service = pin.service
+          type = Datadog::Ext::HTTP::TYPE
+          span = tracer.trace('grape.endpoint_run_filters', service: service, span_type: type)
+
+          # catch thrown exceptions
+          span.set_error(payload[:exception_object]) unless payload[:exception_object].nil?
+
+          span.set_tag('grape.filter.type', type.to_s)
+          span.start_time = start
```
Comment (ufoot) [inline lib/ddtrace/contrib/grape/endpoint.rb]: Same as above, ensure.

File: lib/ddtrace/contrib/grape/patcher.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+      # Patcher that introduces more instrumentation for Grape endpoints, so that
+      # new signals are executed at the beginning of each step (filters, render and run)
+      module Patcher
+        @patched = false
+
+        module_function
+
+        def patch
+          if !@patched && defined?(::Grape)
```
Comment (ufoot) [inline lib/ddtrace/contrib/grape/patcher.rb]: Thanks for using that patcher pattern, :+1:

File: lib/ddtrace/contrib/grape/patcher.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+              Datadog::Contrib::Grape::Endpoint.subscribe()
+            rescue StandardError => e
+              Datadog::Tracer.log.error("Unable to apply Grape integration: #{e}")
+            end
+          end
+          @patched
+        end
+
+        def unpatch
+          # TODO: implement this (revert aliasing)
```
Comment (ufoot) [inline lib/ddtrace/contrib/grape/patcher.rb]: ;)

File: lib/ddtrace/monkey.rb [hunk 5]
```diff
@@ -5,14 +5,21 @@
 # patching code, which is required on demand, when patching.
 require 'ddtrace/contrib/active_record/patcher'
 require 'ddtrace/contrib/elasticsearch/patcher'
-require 'ddtrace/contrib/http/patcher'
+require 'ddtrace/contrib/grape/patcher'
 require 'ddtrace/contrib/redis/patcher'
+require 'ddtrace/contrib/http/patcher'
```
Comment (ufoot) [inline lib/ddtrace/monkey.rb]: Was there an issue with the order? (just curious, I'm not requiring any change here)

File: test/contrib/grape/app.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  namespace :filtered do
+    before do
+      sleep(0.01)
+    end
+
+    after do
+      sleep(0.01)
+    end
+
+    desc 'Returns an error'
```
Comment (ufoot) [inline test/contrib/grape/app.rb]: Is that really the right  comment?

File: lib/ddtrace/contrib/grape/endpoint.rb [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+          ::ActiveSupport::Notifications.subscribe('endpoint_run_filters.grape') do |*args|
+            endpoint_run_filters(*args)
+          end
+        end
+
+        def self.endpoint_start_process(*)
+          return if Thread.current[KEY_RUN]
+
+          # retrieve the tracer from the PIN object
+          pin = Datadog::Pin.get_from(::Grape)
```
Comment (palazzem) [inline lib/ddtrace/contrib/grape/endpoint.rb]: You usually have one Grape for each application (and so process). If this is not true, it means that you're mounting 2 different Grape applications in the same application. If we want to support that, it means that we should even support having 2 Rails applications running together in the same process.

The major issue, is that supporting this is not so easy. We don't have instances here and adding PIN references in all endpoints and subscribers is quite complex I think.

File: lib/ddtrace/monkey.rb (hunk 5 above)
Comment (palazzem) [inline lib/ddtrace/monkey.rb]: oh no, it's just cosmetic :D

File: test/contrib/grape/app.rb (hunk 6 above)
Comment (palazzem) [inline test/contrib/grape/app.rb]: nope ahah

File: lib/ddtrace/contrib/grape/endpoint.rb (hunk 7 above)
Comment (ufoot) [inline lib/ddtrace/contrib/grape/endpoint.rb]: Yup, not worth the complexity of attaching the PIN to instances, I'm just saying -> let's keep aware we have this limitation.

Comment (palazzem) [inline lib/ddtrace/contrib/grape/endpoint.rb]: Yea, considering that the standard case is having one Grape application with namespaces. We still not cover some corner case by the way.

File: lib/ddtrace/contrib/grape/endpoint.rb (hunk 1 above)
Comment (palazzem) [inline lib/ddtrace/contrib/grape/endpoint.rb]: Definitely! forgot to add this enforcing, thanks!

File: test/monkey_test.rb [hunk 8]
```diff
@@ -4,58 +4,69 @@
...
 
     # verify that active_record is not auto patched by default
     Datadog::Monkey.patch_all()
     assert_equal(true, Datadog::Contrib::Elasticsearch::Patcher.patched?)
     assert_equal(true, Datadog::Contrib::HTTP::Patcher.patched?)
     assert_equal(true, Datadog::Contrib::Redis::Patcher.patched?)
+    assert_equal(true, Datadog::Contrib::Grape::Patcher.patched?)
     assert_equal(false, Datadog::Contrib::ActiveRecord::Patcher.patched?)
-    assert_equal({ elasticsearch: true, http: true, redis: true, active_record: false }, Datadog::Monkey.get_patched_modules())
+    assert_equal({ elasticsearch: true, http: true, redis: true, grape: true, active_record: false }, Datadog::Monkey.get_patched_modules())
```
Comment (ufoot) [inline test/monkey_test.rb]: I think we don't run the tests for Grape with old Ruby versions, but I would assume this global monkey patching test is run with oldies such as 1.9. So I'm wondering if we should keep this and just drop it, probably removing the `require 'grape'` in this test would be enough. Let's wait for the CI advice, what I'm saying is: for a first release, if Grape does not work with 1.9, let's save this for later and choose whatever path makes it straight to the release, we don't want to delay 0.7.0 for such a corner case. If all tests pass -> forget this remark.

Comment (palazzem) [inline test/monkey_test.rb]: 👍

**Pull request merged**